addons/stock/wizard/stock_quant_reserve.py

```diff
@@ -0,0 +1,61 @@
+
+    quant_id = fields.Many2one('stock.quant', required=True, readonly=True)
+    location_id = fields.Many2one(related='quant_id.location_id', readonly=True)
+    product_id = fields.Many2one(related='quant_id.product_id', readonly=True)
+    product_uom_id = fields.Many2one(related='product_id.uom_id', readonly=True)
+    lot_id = fields.Many2one(related='quant_id.lot_id', readonly=True)
+    package_id = fields.Many2one(related='quant_id.package_id', readonly=True)
+    available_quantity = fields.Float(related='quant_id.available_quantity', string='Available')
+    qty_to_reserve = fields.Float('To reserve')
+
+    reserve_id = fields.Many2one('stock.quant.reserve')
+
+
+class ReserveStockQuant(models.TransientModel):
+    _name = 'stock.quant.reserve'
+    _description = 'Manually reserve stock'
+
+    move_id = fields.Many2one('stock.move')
+    product_uom_id = fields.Many2one(related='move_id.product_id.uom_id', readonly=True)
+    demand_qty = fields.Float(compute='_compute_demand_qty', store='True')
+    qty_to_reserve = fields.Float('Quantity to reserve', compute='_compute_qty_to_reserve')
+    quant_line_ids = fields.One2many('stock.quant.reserve.line', 'reserve_id', compute='_compute_quant_line_ids', store='True')
+
+    @api.depends('quant_line_ids.qty_to_reserve', 'demand_qty')
+    def _compute_qty_to_reserve(self):
+        for wiz in self:
+            wiz.qty_to_reserve = max(wiz.demand_qty - sum(wiz.quant_line_ids.mapped('qty_to_reserve')), 0)
+
+    @api.depends('move_id')
+    def _compute_demand_qty(self):
+        for wiz in self:
+            wiz.demand_qty = wiz.move_id.product_qty - sum(wiz.move_id.move_line_ids.mapped('reserved_qty'))
+
+    @api.depends('move_id')
+    def _compute_quant_line_ids(self):
+        for wiz in self:
+            move_id = wiz.move_id
+            if not move_id:
+                wiz.quant_line_ids = False
+                continue
+            if wiz.quant_line_ids:
+                continue
+            quant_line_cmds = [Command.clear()]
+            quant_ids = self.env['stock.quant'].search([('product_id', '=', move_id.product_id.id),
+                                                        ('location_id', 'child_of', move_id.location_id.id)])
+            quant_line_cmds += [Command.create({'quant_id': quant.id}) for quant in quant_ids.filtered(lambda q: q.available_quantity > 0)]
+            wiz.quant_line_ids = quant_line_cmds
+
+    def reserve_stock(self):
+        for wiz in self:
+            for line in wiz.quant_line_ids.filtered(lambda l: l.qty_to_reserve > 0):
+                line.quant_id.reserved_quantity += line.qty_to_reserve
+                StockMoveLine = self.env['stock.move.line'].with_context(bypass_reservation_update=True)
+                StockMoveLine.create(wiz.move_id._prepare_move_line_vals(quantity=line.qty_to_reserve, reserved_quant=line.quant_id))
+                wiz.move_id._recompute_state()
```

Comment:
can we move this outside of both of the for loops + batch it with self.move_id._recompute_state()?